### PR TITLE
Support partial update

### PR DESCRIPTION
**Pull request by haolingdong-msft**
This pr fix issue https://github.com/Azure/autorest.java/issues/1050
In this pr, I supported partial update for .java files with `@Generated` annotation, currently they are async and sync clients. We already supported add `@Generated` annotation to members in async and sync clients in this pr: https://github.com/Azure/autorest.java/pull/1239.

I add a flag `partial-update` to enable partial update feature.

### Use Cases

I supported below partial update use cases in this pr:
1. Manually add class member (field / method / constructor) -> keep the added member
2. Manually update method signature, e.g. parameter change, method access level change -> keep the manual changed signature, and not generate the corresponding method with the same method name
3. Manually remove one class member -> if the member's definition is in swagger, this member will be auto generated again
4. Swagger add a new api -> add the new api to generated file
5. Swagger update an existing api -> if the api is auto generated, then the existing generated member will be replaced to the new one. If it is manual updated, we will keep the manual updated member.
6. Swagger delete an existing api -> if the existing api is auto generated, then it should be removed. If it is manual updated, we will keep the manual updated member.
	
### Code Change

The code change is mainly in Postprocessor.java and PartialUpdateHandler.java. It will handle partial update for each .java file after finishing customization. When handling partial update for each file, it will compare **existing file** which has manual update and **generated file** which is generated by autorest. We will keep existing file's format as much as possible.

Detailed logic:
1. Parse existing file content and generated file content using JavaParser
2. Get class members for existing file and generated file
3. Check if the file is in scope of partial update by iterate the members in generated file to see if there is a method has `@Generated` annotation. If it has `@Generated` annotation, then the file is in scope of partial update, otherwise return generated file content directly.
	**A potential issue here**: if existing file generated before we support adding `@Generated` annotation, then the existing file will be considered as all manual code change. So existing file members will be kept. One solution is to add a flag to determine whether we need to do partial update. Another solution is to check if existing file has `@Generated` annotation, but an edge case here is when use manually update all existing file, then existing file does not have `@Generated` annotation, but we still want to handle partial update. Currently I use the first solution to add a flag to enable partial update feature.
4. Iterate existing file members, keep manual updated members, and replace generated members with the corresponding newly generated one. Here we will not do the replace on the existing file member list,  we just create a new member list `updatedMembersList` and put in those manually update members and newly generated members.
5. Add remaining newly generated members to `updatedMembersList`
6. Update generated file members with `updatedMembersList`
7. Update generated file imports

### Unit Test

Unit test is in [PartialUpdateHandlerTest.java](https://github.com/Azure/autorest.java/blob/228cc14d3988b8979c743b23cc198fa27ea4cfa4/postprocessor/src/test/java/com/azure/autorest/postprocessor/util/PartialUpdateHandlerTest.java). it tests below cases:

1. Existing file manually add class member (field / method / constructor)
2. Existing file manually update method signature - parameter change and access level change
3. Existing file manually remove method
4. Generated file add API
5. Generated file remove API
6. Generated file update API
7. Generated file update API and the method is also manually updated in existing file -> keep manual update one
8. Generated file remove API and the method is also manually updated in existing file -> keep manual update one

### Integration Test

I also added integration tests to test the overall process for partial update in partial-update-tests. The test will generate code based on existing folder and [bodystring_updated.json](https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/swagger/specification/bodystring_updated.json). The existing files are in `existing` folder and the generated files are in `generated` folder. `StringOperationClient.java` covers manually code change on existing file, `EnumClient.java` covers swagger change.

Detailed test cases:
1. Manually add class member (field / method / constructor)
* existing code change:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java#L23-L24
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java#L36-L39
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java#L275

* Corresponding generated code: https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/generated/src/main/java/fixtures/bodystring/StringOperationClient.java

2. Manually update method signature - parameter change
* existing code change:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java#L60-L78

* Corresponding generated code: https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/generated/src/main/java/fixtures/bodystring/StringOperationClient.java

3. Manually update method signature -  method accessibility level change
* existing code change:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java#L80-L98

* Corresponding generated code: https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/generated/src/main/java/fixtures/bodystring/StringOperationClient.java

4. Swagger update api - method parameter name change 
*  swagger change:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/swagger/specification/bodystring_updated.json#L299

* Corresponding generated code:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/generated/src/main/java/fixtures/bodystring/EnumClient.java

5. Swagger add new api
* swagger change:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/swagger/specification/bodystring_updated.json#L430

* Corresponding generated code:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/generated/src/main/java/fixtures/bodystring/EnumClient.java

6. Swagger remove api
* swagger change:
Below api from bodystring.json is removed from bodystring_updated.json
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/existing/swagger/specification/bodystring.json#L382-L406

* Corresponding generated code:
https://github.com/Azure/autorest.java/blob/8d8d6d401f387ca085c0681c136da650ed653cfd/partial-update-tests/generated/src/main/java/fixtures/bodystring/EnumClient.java

### CODE REVIEW

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (weidongxu-microsoft) [inline postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:252]: Why not put them to a separate class? At least it will be easier for your unit test later.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (weidongxu-microsoft) [inline postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:144]: You might add a setting to control whether to enable the feature.

File: pom.xml
Comment (weidongxu-microsoft) [inline pom.xml:131]: Add the folder to https://github.com/Azure/autorest.java/blob/main/.npmignore

**Review by weidongxu-microsoft — COMMENTED** [review]
Could you show us where is your customization on the code?

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (haolingdong-msft) [reply postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:252]: Agree, I also plan to do this refactoring, just first check in to show the logic.

**Comment by haolingdong-msft** [comment]
> Could you show us where is your customization (hand-written change) on the test?

Enhanced 'Test' section, please refer to the 'Test' section for my hand-written change

File: partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java
Comment (haolingdong-msft) [inline partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:23]: manually code change is in this file.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (haolingdong-msft) [reply postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:145]: working on adding a flag to enable partial update feature

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (haolingdong-msft) [reply postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:253]: Done  with refactoring

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java
Comment (weidongxu-microsoft) [inline extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java:136]: nit, space

File: generate
Comment (weidongxu-microsoft) [inline generate:97]: suggestion, move `--handle-partial-update` to your swagger/readme?

File: partial-update-tests/existing/swagger/README.md
Comment (weidongxu-microsoft) [inline partial-update-tests/existing/swagger/README.md:16]: these seems not needed, exception `namespace`

File: postprocessor/src/test/resources/partialupdate/StringOperationWithAddedMemberClient.java
Comment (weidongxu-microsoft) [inline postprocessor/src/test/resources/partialupdate/StringOperationWithAddedMemberClient.java:48]: should this be `@Generated`?

File: postprocessor/src/test/resources/partialupdate/StringOperationWithUpdateMemberClient.java
Comment (weidongxu-microsoft) [inline postprocessor/src/test/resources/partialupdate/StringOperationWithUpdateMemberClient.java:41]: Is this for `StringOperationWithUpdateMemberGeneratedClient.java`? If so, should the method be package private (instead of public) in `StringOperationWithUpdateMemberGeneratedClient.java`?

**Review by weidongxu-microsoft — APPROVED** [review]

File: postprocessor/src/test/resources/partialupdate/StringOperationWithUpdateMemberClient.java
Comment (haolingdong-msft) [reply postprocessor/src/test/resources/partialupdate/StringOperationWithUpdateMemberClient.java:41]: `StringOperationWithUpdateMemberGeneratedClient.java`: it simulates swagger update api to add a parameter named `String updateParam` to `putNullWithResponse`. But does not change access level, so access level is still public.
`StringOperationWithUpdateMemberClient.java`: it simulates manual code change to add a parameter named `String test` and modify access level from `public` to `package-private`

File: postprocessor/src/test/resources/partialupdate/StringOperationWithAddedMemberClient.java
Comment (haolingdong-msft) [reply postprocessor/src/test/resources/partialupdate/StringOperationWithAddedMemberClient.java:48]: nice catch, shoud be @Generated,  will update, thanks

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java
Comment (haolingdong-msft) [reply extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java]: updated

File: generate
Comment (haolingdong-msft) [reply generate]: updated

File: partial-update-tests/existing/swagger/README.md
Comment (haolingdong-msft) [reply partial-update-tests/existing/swagger/README.md]: updated

**Comment by haolingdong-msft** [comment]
@srnagar Could you please take a look at this pr and let me know if any comment?

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java
Comment (srnagar) [inline extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java:136]: nit: I think we can drop `handle` from the name. Just `partial-update` should be sufficient.

File: partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java
Comment (srnagar) [inline partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:275]: Can you also add an annotation here and also include JavaDocs to ensure, it doesn't get overwritten?

File: partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java
Comment (srnagar) [inline partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:292]: Are we using `@Generated` annotation for `Client` classes only? What about builders? We may manually add some builder options that don't get generated. For e.g. some services support `connectionString` which we don't generate but could be added manually.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java
Comment (srnagar) [inline postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java:16]: The documentation you have in the PR description is good. We should include that as JavaDoc for this class.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (srnagar) [inline postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:262]: If there was an error, we should throw the exception and not continue as this would overwrite an existing file that may have required a partial update.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java
Comment (srnagar) [inline postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java:46]: We should call this list `manuallyAddedMembersList` or something similar to make it explicit that these are the members that were not generated.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (srnagar) [inline postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:256]: Why do we need to get the URI? Can we just do the following?
```java
if (Files.exists(Paths.get(projectBaseDirectoryPath, path))) 
```

File: postprocessor/src/test/resources/partialupdate/StringOperationWithRemovedMemberGeneratedClient.java
Comment (srnagar) [inline postprocessor/src/test/resources/partialupdate/StringOperationWithRemovedMemberGeneratedClient.java:20]: nit: it would be better to have `@Generated` in a separate line. It's the pattern we use for `@JsonProperty` annotation too.

File: partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java
Comment (haolingdong-msft) [reply partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:292]: Currently, we only include `@Generated` for `Client` classes. We can extend to Builders later.

Comment (srnagar) [reply partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:292]: I guess the builder, sync and async types should have `@Generated` annotations right from the beginning as those are the most likely user-facing types that will be manually edited.

Comment (haolingdong-msft) [reply partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:292]: OK, thanks for pointing out, in this issue https://github.com/Azure/autorest.java/issues/1099, we noted down that we will support only for clients as first step, so I did not add to builders. But I will create a issue to add `@Generated` annotation to builders and work on that as my next task. 
Do you think we should block this pr merge until I finish adding the annotation to builders, or we can merge this pr first ? ( I will modify this pr according to the comment and let you review again before merging)

Comment (haolingdong-msft) [reply partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:292]: issue created: https://github.com/Azure/azure-sdk-for-java/issues/25895

Comment (weidongxu-microsoft) [reply partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:292]: At present dev can use `--disable-client-builder` to avoid generate the Builder, if they've done customization. Builder itself is not commonly updated during swagger update.

Handling of Builder will be handled in a next PR. We do not want to have a super big PR for everything.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java
Comment (haolingdong-msft) [reply postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java:46]: Because `updatedMemberList` is the member list that we will set to `generatedClazz`, 
```java
generatedClazz.setMembers(updatedMembersList);
``` 
It not only contains manually added member, but also contains generated members, 'Code Change' section in description explains more details about `updatedMemberList`

File: postprocessor/src/test/resources/partialupdate/StringOperationWithRemovedMemberGeneratedClient.java
Comment (haolingdong-msft) [reply postprocessor/src/test/resources/partialupdate/StringOperationWithRemovedMemberGeneratedClient.java:20]: In `ServiceSyncClientTemplate` (https://github.com/haolingdong-msft/autorest.java/blob/e4405a9c646d4a1306b9fb0eafd58049dfd42925/javagen/src/main/java/com/azure/autorest/template/ServiceSyncClientTemplate.java), we add `@Generated` for member variables **in a seperate line**. But after formatting, `@Generated` comes to the same line, it only happens to member variable declaration. So may need more investigation on 
 the formatter if we want `@Generated` to be in a seperate line.
```java
// Add service client member variable
      classBlock.annotation("Generated"); // This hold a seperate line
      if (wrapServiceClient) {
        classBlock.privateFinalMemberVariable(serviceClient.getClassName(), "serviceClient");
      } else {
        classBlock.privateFinalMemberVariable(methodGroupClient.getClassName(), "serviceClient");
      }
```

Comment (weidongxu-microsoft) [reply postprocessor/src/test/resources/partialupdate/StringOperationWithRemovedMemberGeneratedClient.java:20]: @srnagar This is due to your javaformat. If you want this format, please update it.

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (haolingdong-msft) [reply postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java:262]: updated

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java
Comment (haolingdong-msft) [reply postprocessor/src/main/java/com/azure/autorest/postprocessor/util/PartialUpdateHandler.java:30]: updated

File: postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java
Comment (haolingdong-msft) [reply postprocessor/src/main/java/com/azure/autorest/postprocessor/Postprocessor.java]: updated

File: partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java
Comment (haolingdong-msft) [reply partial-update-tests/existing/src/main/java/fixtures/bodystring/StringOperationClient.java:281]: updated

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java
Comment (haolingdong-msft) [reply extension-base/src/main/java/com/azure/autorest/extension/base/plugin/JavaSettings.java]: updated

**Pull request merged**
